source/postprocess/crystal_preferred_orientation.cc

```diff
@@ -0,0 +1,816 @@
+
+      const auto &particle_handler = this->get_particle_world().get_particle_handler();
+
+      std::stringstream string_stream_master;
+      std::stringstream string_stream_content_raw;
+      std::stringstream string_stream_content_draw_volume_weighting;
+
+      string_stream_master << "id x y" << (dim == 3 ? " z" : " ") << " olivine_deformation_type" << std::endl;
+
+      // get particle data
+      bool wrote_weighted_header = false;
+      bool wrote_unweighted_header = false;
+      for (typename dealii::Particles::ParticleHandler<dim>::particle_iterator it = particle_handler.begin(); it != particle_handler.end(); ++it)
+        {
+
+          AssertThrow(it->has_properties(),
+                      ExcMessage("No particle properties found. Make sure that the CPO particle property plugin is selected."));
+
+
+
+          unsigned int id = it->get_id();
+          const ArrayView<double> &properties = it->get_properties();
+
+          const Particle::Property::ParticlePropertyInformation &property_information = this->get_particle_world().get_property_manager().get_data_info();
+
+          AssertThrow(property_information.fieldname_exists("cpo mineral 0 type") ,
+                      ExcMessage("No CPO particle properties found. Make sure that the CPO particle property plugin is selected."));
+
+
+
+          const unsigned int cpo_data_position = property_information.n_fields() == 0
+                                                 ?
+                                                 0
+                                                 :
+                                                 property_information.get_position_by_field_name("cpo mineral 0 type");
+
+          Point<dim> position = it->get_location();
+
+          // always make a vector of rotation matrices
+          std::vector<std::vector<Tensor<2,3>>> rotation_matrices;
+          rotation_matrices.resize(n_minerals);
+          for (unsigned int mineral_i = 0; mineral_i < n_minerals; ++mineral_i)
+            {
+              rotation_matrices[mineral_i].resize(n_grains);
+              for (unsigned int i_grain = 0; i_grain < n_grains; i_grain++)
+                {
+                  rotation_matrices[mineral_i][i_grain] = cpo_particle_property.get_rotation_matrix_grains(
+                                                            cpo_data_position,
+                                                            properties,
+                                                            mineral_i,
+                                                            i_grain);
+                }
+            }
+          // write master file
+          string_stream_master << id << " " << position << " " << properties[cpo_data_position] <<  std::endl;
+
+          // write content file
+          std::vector<std::vector<std::vector<double>>> euler_angles;
```

Comment:
same as above, move outside loop, and:
```suggestion
          std::vector<std::vector<std::vector<double>>> euler_angles(n_minerals,{n_grains});
```